### PR TITLE
[stable/nextcloud] Add prometheus metrics exporter

### PR DESCRIPTION
**Pull request by tomaaron**
Signed-off-by: Tom Koch <tomaaronkoch@gmail.com>

#### Is this a new chart
no

#### What this PR does / why we need it:
This affects only **Nextcloud** chart.

This PR introduces the possibility to add a single Prometheus metrics exporter using the following [xperimental/nextcloud-exporter](https://github.com/xperimental/nextcloud-exporter). The metrics exporter is disabled by default.
#### Which issue this PR fixes
none
#### Special notes for your reviewer:
none
#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[stable/mychartname]`)

### CODE REVIEW

**Comment by tomaaron** [comment]
/assign @billimek
I've locally linted the chart and deployed it to a K8s testcluster but I can't see why the ci/cirleci: lint-chart fails.

**Comment by tomaaron** [comment]
Ah, got it! Missing new line is fixed.

**Comment by mrueg** [comment]
@tomaaron thanks was looking forward to seeing that! Can you add a ServiceMonitor object as well for Prometheus-Operator?

**Comment by tomaaron** [comment]
> @tomaaron thanks was looking forward to seeing that! Can you add a ServiceMonitor object as well for Prometheus-Operator?

Bummer! Right now I've no setup with a Prometheus-Operator and therefore I couldn't test a possible ServiceMonitor.

File: stable/nextcloud/values.yaml [hunk 1]
```diff
@@ -345,3 +345,30 @@ nodeSelector: {}
...
+metrics:
+  enabled: false
+
+  # The metrics exporter needs to know how you serve Nextcloud either http or https
+  https: false
+  timeout: 5s
+
+  image:
+    repository: xperimental/nextcloud-exporter
+    tag: latest
```
Comment (mrueg) [inline stable/nextcloud/values.yaml]: would suggest to use v0.3.0 here to be able to reproduce issues if they occur.

File: stable/nextcloud/templates/metrics-deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+            secretKeyRef:
+              name: {{ template "nextcloud.fullname" . }}
+              key: nextcloud-password
+        - name: NEXTCLOUD_SERVER
+          value: http{{ if .Values.metrics.https }}s{{ end }}://{{ .Values.nextcloud.host }}
+        - name: NEXTCLOUD_TIMEOUT
+          value: {{ .Values.metrics.timeout }}
+        ports:
+        - name: metrics
+          containerPort: 9205
```
Comment (mrueg) [inline stable/nextcloud/templates/metrics-deployment.yaml:50]: you could expose them via service object

**Comment by tomaaron** [comment]
Hey Manual, thanks for the recommendations. I've added a service for the metrics and fixed the image tag.

File: stable/nextcloud/README.md [hunk 3]
```diff
@@ -150,6 +150,14 @@ The following table lists the configurable parameters of the nextcloud chart and
...
+| `metrics.https`                                              | Wether https is used for scraping Nextcloud metrics     | `false`                                     |
+| `metrics.timeout`                                            | When the scrape times out                               | `5s`                                        |
+| `metrics.image.repository`                                   | Nextcloud metrics exporter image name                   | `xperimental/nextcloud-exporter`            |
+| `metrics.image.tag`                                          | Nextcloud metrics exporter image tag                    | `latest`                                    |
+| `metrics.image.pullPolicy`                                   | Nextcloud metrics exporter image pull policy            | `IfNotPresent`                              |
+| `metrics.podAnnotations`                                     | Additional annotations for metrics exporter             | `{prometheus.io/scrape: "true", prometheus.io/port: "9205"}` |
```
Comment (mrueg) [inline stable/nextcloud/README.md]: these would probably then be rather annotations for the service (if you expose via a service)

File: stable/nextcloud/templates/metrics-deployment.yaml [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+metadata:
+  name: {{ template "nextcloud.fullname" . }}-metrics
+  labels:
+    app.kubernetes.io/name: {{ include "nextcloud.name" . }}
+    helm.sh/chart: {{ include "nextcloud.chart" . }}
+    app.kubernetes.io/instance: {{ .Release.Name }}
+    app.kubernetes.io/managed-by: {{ .Release.Service }}
+    app.kubernetes.io/component: metrics
+spec:
+  replicas: 1
```
Comment (mrueg) [inline stable/nextcloud/templates/metrics-deployment.yaml]: Can you make the number of replicas configurable as well?

File: stable/nextcloud/README.md [hunk 5]
```diff
@@ -150,6 +150,14 @@ The following table lists the configurable parameters of the nextcloud chart and
 | `hpa.maxPods`                                                | Max. pods for the Nextcloud HorizontalPodAutoscaler     | `10`                                        |
 | `deploymentAnnotations`                                      | Annotations to be added at 'deployment' level           | not set                                     |
 | `podAnnotations`                                             | Annotations to be added at 'pod' level                  | not set                                     |
+| `metrics.enabled`                                            | Start a single Prometheus metrics exporter              | `false`                                     |
+| `metrics.https`                                              | Wether https is used for scraping Nextcloud metrics     | `false`                                     |
```
Comment (mrueg) [inline stable/nextcloud/README.md]: This is a bit misleading (at least to me)? The value sounds like you enable https on the exporter
While this defines if https or http is used to connect to nextcloud, right?

Comment (tomaaron) [reply stable/nextcloud/README.md]: Yes, you're right it's about the connection between nextcloud and the exporter. The exporter unfortunately needs this explicitly. I'm also not so happy with the naming. Do you have a suggestion for the name?

File: stable/nextcloud/README.md (hunk 3 above)
Comment (tomaaron) [reply stable/nextcloud/README.md]: I've moved the annotations to the service object.

File: stable/nextcloud/templates/metrics-deployment.yaml (hunk 4 above)
Comment (tomaaron) [reply stable/nextcloud/templates/metrics-deployment.yaml]: Sure thing!

File: stable/nextcloud/README.md [hunk 6]
```diff
@@ -150,14 +150,19 @@ The following table lists the configurable parameters of the nextcloud chart and
 | `hpa.maxPods`                                                | Max. pods for the Nextcloud HorizontalPodAutoscaler     | `10`                                        |
 | `deploymentAnnotations`                                      | Annotations to be added at 'deployment' level           | not set                                     |
 | `podAnnotations`                                             | Annotations to be added at 'pod' level                  | not set                                     |
-| `metrics.enabled`                                            | Start a single Prometheus metrics exporter              | `false`                                     |
-| `metrics.https`                                              | Wether https is used for scraping Nextcloud metrics     | `false`                                     |
+| `metrics.enabled`                                            | Start Prometheus metrics exporter                       | `false`                                     |
+| `metrics.https`                                              | Defines if https is used to connect to nextcloud        | `false` (uses http)                         |
```
Comment (tomaaron) [inline stable/nextcloud/README.md:154]: @mrueg Does this edit of the documentation improves the clarity?

Comment (mrueg) [reply stable/nextcloud/README.md:154]: Yup, thanks!

**Comment by mrueg** [comment]
looks good to me. I can't approve or lgtm, so leaving that to @billimek

**Comment by chrisingenhaag** [comment]
Hey @tomaaron ,
thx for the PR. Really nice improvement.

/lgtm

**Pull request merged**